Index: linux-2.6.24.3/include/linux/Kbuild
===================================================================
--- linux-2.6.24.3.orig/include/linux/Kbuild	2008-06-03 13:06:01.000000000 +0200
+++ linux-2.6.24.3/include/linux/Kbuild	2008-06-03 13:06:04.000000000 +0200
@@ -217,7 +217,6 @@
 unifdef-y += icmp.h
 unifdef-y += icmpv6.h
 unifdef-y += if_addr.h
-unifdef-y += if_addrlabel.h
 unifdef-y += if_arp.h
 unifdef-y += if_bridge.h
 unifdef-y += if_ec.h
