Index: qtopia-core-opensource-src-4.4.0-snapshot-20080212/src/corelib/global/qglobal.h
===================================================================
--- qtopia-core-opensource-src-4.4.0-snapshot-20080212.orig/src/corelib/global/qglobal.h	2008-02-19 09:58:52.000000000 +0100
+++ qtopia-core-opensource-src-4.4.0-snapshot-20080212/src/corelib/global/qglobal.h	2008-02-19 09:59:05.000000000 +0100
@@ -951,7 +951,7 @@
 
 typedef int QNoImplicitBoolCast;
 
-#if defined(QT_ARCH_ARM) || (defined(QT_ARCH_MIPS) && (defined(Q_WS_QWS) || defined(Q_OS_WINCE)))
+#if defined(QT_ARCH_ARM) || defined(QT_ARCH_AVR32) || (defined(QT_ARCH_MIPS) && (defined(Q_WS_QWS) || defined(Q_OS_WINCE)))
 #define QT_NO_FPU
 #endif
 
