--- qtopia-core-opensource-src-4.4.0-snapshot-20080120/src/gui/embedded/qmousetslib_qws.cpp	2008-01-25 10:39:45.000000000 +0100
+++ qtopia-core-opensource-src-4.4.0-snapshot-20080120/src/gui/embedded/qmousetslib_qws.new.cpp	2008-01-25 10:40:04.000000000 +0100
@@ -61,6 +61,10 @@ QT_BEGIN_NAMESPACE
 #define QT_QWS_TP_JITTER_LIMIT 3
 #endif
 
+#ifdef TSLIBMOUSEHANDLER_DEBUG
+#include <QDebug>
+#endif
+
 /*!
     \internal
 
